File: golang/wiggle_subsequence.go
package main

func wiggleMaxLength(nums []int) int {
	n := len(nums)
	if n == 1 {
		return 1
	}

	var disc []int
	for i := 1; i < n; i++ {
		disc = append(disc, nums[i]-nums[i-1])
	}

	ans, last := 0, 0
	for i := 0; i < len(disc); i++ {
		if (last > 0 && disc[i] > 0) || (last < 0 && disc[i] < 0) || (disc[i] == 0) {
			continue
		}

		ans++
		last = disc[i]
	}

	return ans + 1
}

File: python3/wiggle_subsequence.py
"""
Space   : O(n)
Time    : O(n)
"""


class Solution:
    def wiggleMaxLength(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return 1

        disc = []
        for i in range(1, n):
            disc.append(nums[i] - nums[i-1])

        ans = last = 0
        for i, v in enumerate(disc):
            if (last > 0 and v > 0) or (last < 0 and v < 0) or (v == 0):
                continue

            ans += 1
            last = v

        return ans + 1
